### DISCUSSION TITLE
Weird Z-Offset issue

### DISCUSSION THREAD

**Discussion by tinooo** [Jyers/Marlin#173] [Q&A] [answered]
Hi,

I'm not sure what I'm doing wrong here or if it actually is a bug (but i don't think so).
I've a capacitive ABL Sensor on my Ender 3v2. I've compiled my own firmware. (Config.h attached).

Basically the ABL semms to work. G29 and it probes everything, checking with M420 the mesh seems also to be quite good. Repeatability test is also very acurate.

When having my printer run several prints, the Z-Offset seems to change quite drastically. Here is what I do:

1. Heat Bed
2. Run Mesh leveling and store to eeprom
3. Heat Bed&Nozzle
4. Run "Live Adjust" Z-Offset and use paper to determine Z-Offset --> approx.: -2.3mm
5. Save data.
6. Run Print with G28 and M420 S1
7. Result seems to be quite good. Maybe adjust Z-offset slightly using "fine Tune" ±0.05mm for perfect layer
8. Print finishes after ~1-2h
9. Start next print (G28, M420 S1)
10. Again have to adjust Z-offset slightly.
11. Repeat Steps 8-10 a few times (3-4 Prints)
12. Now it is possible that my Z-Offset is -2.7mm (note initial offset was set to -2.3mm to get good results)

How can this be? I'm aware, that a captive sensor is effected by humidity and temperature. But I don't think this can be the reason. I print in a heated room, ~21°C. During the process, temperature increases max 2°C due to printer heat and my emitted heat :p 
Humidity also might increase slightly, but imho this can't be the reason for this huge change. Also I make sure that the bed is heated, each time before Homing.

Screws of the Z-Probe mount seem also to be tight. Also have checkt the rest of the printer for lose parts.

Any ideas?

Thanks in advance!!

Configuration.h:
[Configuration.txt](https://github.com/Jyers/Marlin/files/6081991/Configuration.txt)

**Comment by Schnabulation** [comment, marked as answer]
I think (but it's just a shot in the dark) the fine tune you are using in step 7 does not get written to the EEPROM. I have noticed that the mesh data is quite dependent on the settings in EEPROM. So when you adjust the Z-offset and do not write it to EEPROM (either because you forget it or you do it while printing, which - as said above - seems not to get written to EEPROM), it will not be available next print.

I even found that when you tweak the Z-offset with the Z-offset menu and "Live adjust" you HAVE to run a new G29, before the new settings get applied, even when you store them to EEPROM.

Personally I tweak the Z-offset by using "Live adjust" and then I transfer it to configuration.h and recompile the firmware. Like this it's always on the correct value and I don't have to mess with it.

**Reply by tinooo** [reply]
shit, that would explain quite a lot! I've spent hours, trying to relevel, sending manual gCodes and doing test prints. I havn't been able to find any pattern in the problems I've had.

But that would explain plenty of my issues. The Display is geeting updated with the "live adjust data" and it getting somhow stored. and on the next print it's starting with the eeprom value again (but the display still showing the last values). So i will gradually increase the display, but in fact the "real" offset is not being increased that much.

Also it somehow would match with my experience, that I've had better/more consistent prints/Zoffsets, when using G29 each time, instead of M420. Since if it is like you say, when altering the Z-Offset (and stroing it to eeprom) without a G29 it also is not beeing taken into account properly and, again I've to drastically change the live adjust during start of the print.

I'll follow your advice and put the offset directly in my config.h, after finding the sweetspot.

much appreciated!


Edit: I just realized, that also would explain my nozzle ramming into the bed.
after some prints, gradually increasing (... achtually reducing) the z-Offset from from the live adjust during print, I went to the seetings menu, adapting the value from the Z-offset there (e.g. -2.3mm to -2.7mm) and storing it in the Eeprom.
In the next print, my nozzle was grinding on the glass, making it almost impossible removing the first layer. This again could be explained by the behavior you mentioned...

I'll check over the weekend and come back with the results. But think that is quite promising. Hope this thread will also help future people

**Reply by Schnabulation** [reply]
Hey nice, glad I could help!

Personally I don't really adjust the Z-offset all that much therefore I just stored it in my configuration.h and leave it as is. Also I re-level with G29 before every print - yes, takes some time but with a 3x3 grid it's not that long and I had great success. I actually have no adhesion problems whatsoever - very reliable like this.

EDIT: https://marlinfw.org/docs/gcode/M290.html
"Note that when BABYSTEP_ZPROBE_OFFSET is enabled, M290 also modifies the Probe Z Offset (with no immediate effects). The new Z offset applies to successive probing operations, and can be saved with M500"
--> I think that's why the Z-offset is only applied after a new G29.

**Reply by tinooo** [reply]
Unfortunately I've quite a warped bed and use 9x9 probes (even thought about trying to go higher). And since I've done plenty of small prints for my drone (1-3g plastic max), the leveling each time was quite annoying. That's when I went with M420 S1.

**Reply by Schnabulation** [reply]
There is a pull request in the making for enabling unified bed leveling: https://marlinfw.org/docs/features/unified_bed_leveling.html
Basically you do a very high resolution probe (ex. 15x15 points) and store this mesh in EEPROM. Then, before every print, you only do a 2x2 probe to compensate for any tilt you might have gotten in the meantime.

It's pretty awesome from what I've read but I haven't used it myself.
